### ISSUE TITLE
signals: "file_downloaded" sent too early

### ISSUE THREAD

**Issue by slint** [inveniosoftware/invenio-files-rest#180]
The `file_downloaded` signal [is sent too early during the file download request](https://github.com/inveniosoftware/invenio-files-rest/blob/94214019b8b7405067555cb34ab11f051a28bb3d/invenio_files_rest/views.py#L597). This is bad for e.g. `invenio-stats`, which relies on the signal for gathering file download statistics. In case the download is interrupted for some reason (e.g. connection dropped or the user canceled the download), the signal will already have been sent.

A possible solution is to use the [`flask.after_this_request` callback](http://flask.pocoo.org/docs/1.0/patterns/deferredcallbacks/), and send the signal afterward if the request was successful (i.e. 2xx/3xx).

**Comment by lnielsen** [comment]
In case ``send_file()`` sends a redirect (e.g. x-sendfile) won't this be the same problem?

**Comment by slint** [comment]
True. In general, after some local testing, apparently, it's not easy (if even possible at all) to keep track of when a file was actually downloaded or if the connection was dropped... One idea is using [`werkzeug.wsgi.ClosingIterator`](https://github.com/pallets/werkzeug/blob/d129d17066768238139b8f8a5d667d9a4f4d68db/werkzeug/wsgi.py#L841) which can execute a callback after the `.close()` is called for the file, but this still doesn't solve the issue...

*(TBH the issue rises from what COUNTER and other discussions related to datasets/files would like to track about files. One of the things e.g. is the total transferred volume of a dataset/file, which one cannot know unless we know that the file arrived someplace...)*
